exonum-java-binding/core/src/test/java/com/exonum/binding/core/runtime/TransactionMethodExtractorTest.java

```diff
@@ -0,0 +1,185 @@
+            + ". But second parameter type was: " + byte[].class.getName(),
+        methodName, DuplicateTransactionMethodArgumentsService.class.getName());
+    assertThat(e.getMessage()).contains(errorMessage);
+  }
+
+  @Test
+  void findMethodsValidServiceInterfaceImplementation() throws Exception {
+    Map<Integer, Method> transactions =
+        TransactionMethodExtractor.findTransactionMethods(
+            ValidServiceInterfaceImplementation.class);
+    assertThat(transactions).hasSize(2);
+    Method transactionMethod = ValidServiceInterfaceImplementation.class.getMethod(
+        "transactionMethod", byte[].class, TransactionContext.class);
+    Method transactionMethod2 = ValidServiceInterfaceImplementation.class.getMethod(
+        "transactionMethod2", byte[].class, TransactionContext.class);
+    List<Method> actualMethods = Arrays.asList(transactionMethod, transactionMethod2);
+    assertThat(actualMethods).containsExactlyInAnyOrderElementsOf(transactions.values());
+  }
+
+  class BasicService implements Service {
+
+    static final int TRANSACTION_ID = 1;
+
+    @Override
+    public List<HashCode> getStateHashes(Snapshot snapshot) {
+      return Collections.emptyList();
+    }
+
+    @Override
+    public void createPublicApiHandlers(Node node, Router router) {
+      // no-op
+    }
+  }
+
+  class ValidService extends BasicService {
+
+    @TransactionMethod(TRANSACTION_ID)
+    @SuppressWarnings("WeakerAccess") // Should be accessible
+    public void transactionMethod(byte[] arguments, TransactionContext context) {}
+  }
+
+  class DuplicateTransactionIdsService extends BasicService {
+
+    @TransactionMethod(TRANSACTION_ID)
+    public void transactionMethod(byte[] arguments, TransactionContext context) {}
+
+    @TransactionMethod(TRANSACTION_ID)
+    public void anotherTransactionMethod(byte[] arguments, TransactionContext context) {}
+  }
+
+  class MissingTransactionMethodArgumentsService extends BasicService {
+
+    @TransactionMethod(TRANSACTION_ID)
+    public void transactionMethod(byte[] arguments) {}
+  }
+
+  class InvalidTransactionMethodArgumentsService extends BasicService {
+
+    @TransactionMethod(TRANSACTION_ID)
+    public void transactionMethod(byte[] arguments, String invalidArgument) {}
+  }
+
+  class DuplicateTransactionMethodArgumentsService extends BasicService {
+
+    @TransactionMethod(TRANSACTION_ID)
+    public void transactionMethod(byte[] arguments, byte[] context) {}
+  }
+
+  interface ServiceInterface {
+    int TRANSACTION_ID = 1;
+
+    @TransactionMethod(TRANSACTION_ID)
+    void transactionMethod(byte[] arguments, TransactionContext context);
+  }
+
+  class ValidServiceInterfaceImplementation implements ServiceInterface {
+
+    static final int TRANSACTION_ID_2 = 2;
+
+    @TransactionMethod(TRANSACTION_ID)
+    public void transactionMethod(byte[] arguments, TransactionContext context) {}
```

Comment:
I considered this case (the present *duplicates* the annotation which isn't something that shall happen, I think):
```java
  interface ServiceInterface {

    @TransactionMethod(1) // <-- Annotation is present on the interface
    void transactionMethod(byte[] arguments, TransactionContext context);
  }

  class ServiceInterfaceImpl implements ServiceInterface {

    @Override // <-- NO duplicate TransactionMethod annotation!
    public void transactionMethod(byte[] arguments, TransactionContext context) {}

  }
```

I think it is better to offload this to a sep. issue because there are more to that: e.g., what if an interface method and the implementation method have *different* @TransactionMethods?

Please document the considerations in a new issue.

Reply:
Created ECR-3861